sphinx/application.py

```diff
@@ -338,6 +338,17 @@ def build(self, force_all=False, filenames=None):
                                  (status, self._warncount)))
             else:
                 logger.info(bold(__('build %s.') % status))
+
+            if self.statuscode == 0 and self.builder.epilog:
```

Comment:
What do you think about moving this to the end of `Builder.build()`? I think emitting epilog message is responsibility of builder itself.

In addition, I'm worried about `epilog` is better way or not. I think a method of builder is also good like `Builder.show_finish_message()`.

Reply:
> What do you think about moving this to the end of `Builder.build()`? I think emitting epilog message is responsibility of builder itself.

I can do this. The reason I kept it separate was I wasn't sure if `Builder.build()` was ever overridden by subclasses. If it was, we could end up emitting the message before any subclasses were called. For example:

    class MySpecialBuilder(Builder):
        ...
        def build(self, docnames, summary=None, method='update'):
            super(MySpecialBuilder, self).build(docnames, summary, method)
            # do something else here

It's your call though.

> In addition, I'm worried about epilog is better way or not. I think a method of builder is also good like `Builder.show_finish_message()`.

I prototyped with both a function and a property but both seemed overkill. The logic is always the same: if the build succeeds then emit the message. Using `printf` formatting will also ensure we get consistent builder messages. Is there any reason _not_ to do things this way?

Reply:
> It's your call though. Let me know what you'd prefer.

Actually, there was another reason. I wanted the summary message to come *after* the `build succeeded` message. For example:

    ...
    build succeeded

    You can now process the pickle files in build/pickle.

I wasn't able to do this if I called this as part of `Builder.finish()`. If we move this here, we should also move the generation of the `build succeeded` message. I don't know if we want to do this.

Reply:
> If it was, we could end up emitting the message before any subclasses were called.
...
>Actually, there was another reason. I wanted the summary message to come after the build succeeded message. 

Reasonable. My idea was wrong. It should be shown on `Builder.build()`.
Please forget this comment.

> Is there any reason not to do things this way?

My worry is how to customize the epilog message on subclasses.
If they want to show more complicated message, an attribute is not extensible.
For example, it is hard to use `out_suffix` to the message. On the other hand, a method is easy to override in subclasses.

I can't determine the interface is important or not. So I'd like to hear your idea.

Reply:
> > Is there any reason not to do things this way?
> 
> My worry is how to customize the epilog message on subclasses.
> If they want to show more complicated message, an attribute is not extensible.
> For example, it is hard to use `out_suffix` to the message. On the other hand, a method is easy to override in subclasses.
> 
> I can't determine the interface is important or not. So I'd like to hear your idea.

Sorry for the delay. The `epilog` attribute covers all the provided builders. Also, if users _really_ need to override this, then can use properties.

    @property
    def epilog(self):
        ...

Personally, I think this is adequate.